File: test/warnings/array_types_warning/python/PackedArrayUnpackableExternElementTest.py
import os
import zserio

import ArrayTypesWarning

from testutils import getApiDir

class PackedArrayUnpackableExternElementTest(ArrayTypesWarning.TestCase):
    def testWriteRead(self):
        packedArrayUnpackableExternElement = self.api.PackedArrayUnpackableExternElement(
            [10, 11, 12],
            [
                zserio.BitBuffer(bytes([0xff, 0xc0]), 10),
                zserio.BitBuffer(bytes([0xff, 0x80]), 10),
                zserio.BitBuffer(bytes([0xff, 0x40]), 10)
            ]
        )

        zserio.serialize_to_file(packedArrayUnpackableExternElement, self.BLOB_NAME)
        readPackedArrayUnpackableExternElement = zserio.deserialize_from_file(
            self.api.PackedArrayUnpackableExternElement, self.BLOB_NAME
        )
        self.assertEqual(packedArrayUnpackableExternElement, readPackedArrayUnpackableExternElement)

    BLOB_NAME = os.path.join(getApiDir(os.path.dirname(__file__)),
                             "packed_array_unpackable_extern_element.blob")

File: test/warnings/array_types_warning/cpp/PackedArrayUnpackableExternElementTest.cpp
#include "gtest/gtest.h"

#include "array_types_warning/packed_array_unpackable_extern_element/PackedArrayUnpackableExternElement.h"

#include "zserio/SerializeUtil.h"

namespace array_types_warning
{
namespace packed_array_unpackable_extern_element
{

using allocator_type = PackedArrayUnpackableExternElement::allocator_type;
template <typename T>
using vector_type = zserio::vector<T, allocator_type>;
using BitBuffer = zserio::BasicBitBuffer<allocator_type>;

class PackedArrayUnpackableExternElementTest : public ::testing::Test
{
protected:
    static const std::string BLOB_NAME;
};

const std::string PackedArrayUnpackableExternElementTest::BLOB_NAME =
        "warnings/array_types_warning/packed_array_unpackable_extern_element.blob";

TEST_F(PackedArrayUnpackableExternElementTest, writeRead)
{
    PackedArrayUnpackableExternElement packedArrayUnpackableExternElement(
        {{10, 11, 12}},
        {{
            BitBuffer(vector_type<uint8_t>{{0xFF, 0xC0}}, 10),
            BitBuffer(vector_type<uint8_t>{{0xFF, 0x80}}, 10),
            BitBuffer(vector_type<uint8_t>{{0xFF, 0x40}}, 10)
        }}
    );

    zserio::serializeToFile(packedArrayUnpackableExternElement, BLOB_NAME);
    auto readPackedArrayUnpackableExternElement =
            zserio::deserializeFromFile<PackedArrayUnpackableExternElement>(BLOB_NAME);
    ASSERT_EQ(packedArrayUnpackableExternElement, readPackedArrayUnpackableExternElement);
}

} // namespace packed_array_unpackable_extern_element
} // namespace array_types_warning
